symengine/sets.cpp

```diff
@@ -804,4 +905,48 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
     return container->set_complement(universe);
 }
 
+RCP<const Set> conditionset(const RCP<const Basic> &sym,
+                            const RCP<const Boolean> &condition)
+{
+    if (eq(*condition, *boolean(false))) {
+        return emptyset();
+    } else if (eq(*condition, *boolean(true))) {
+        return universalset();
+    }
+    if (is_a<And>(*condition)) {
+        auto cont = down_cast<const And &>(*condition).get_container();
+        set_boolean newcont;
+        set_basic present, others;
+        for (auto it = cont.begin(); it != cont.end(); it++) {
+            if (is_a<Contains>(**it)
+                and eq(*down_cast<const Contains &>(**it).get_expr(), *sym)
+                and is_a<FiniteSet>(
+                        *down_cast<const Contains &>(**it).get_set())) {
+                auto fset = down_cast<const Contains &>(**it).get_set();
+                auto fcont
+                    = down_cast<const FiniteSet &>(*fset).get_container();
+                // use the result of simplification done in `logical_and()`
+                for (const auto &elem : fcont) {
+                    if (not(is_a_Number(*elem) or is_a<Constant>(*elem))) {
+                        others.insert(elem);
+                    } else {
+                        present.insert(elem);
+                    }
+                }
+            } else {
+                newcont.insert(*it);
+            }
+        }
+        if (not present.empty()) {
+            newcont.insert(finiteset(others)->contains(sym));
+            return SymEngine::set_union(
+                {finiteset(present), conditionset(sym, logical_and(newcont))});
```

Comment:
What's the result of `And(Contains(x, {1, 2, 3}), Eq(y, x))` and `conditionset(x, And(Contains(x, {1, 2, 3}), Eq(y, x)))`?